src/decode.rs

```diff
@@ -3806,1159 +3806,204 @@ unsafe fn decode_b(
         case_set(bh4, &mut t.l, 1, by4 as isize, &mut set_ctx);
         case_set(bw4, &mut *t.a, 0, bx4 as isize, &mut set_ctx);
 
-        if b.c2rust_unnamed.c2rust_unnamed.pal_sz[0] != 0 {
-            let pal: *mut uint16_t = if t.frame_thread.pass != 0 {
-                ((*(f.frame_thread.pal).offset(
-                    (((t.by >> 1) + (t.bx & 1)) as isize * (f.b4_stride >> 1)
-                        + ((t.bx >> 1) + (t.by & 1)) as isize) as isize,
-                ))[0])
-                    .as_mut_ptr()
+        if b.pal_sz()[0] != 0 {
+            let pal = if t.frame_thread.pass != 0 {
+                let index = ((t.by >> 1) + (t.bx & 1)) as isize * (f.b4_stride >> 1)
+                    + ((t.bx >> 1) + (t.by & 1)) as isize;
+                &(*f.frame_thread.pal.offset(index))[0]
             } else {
-                (t.scratch.c2rust_unnamed_0.pal[0]).as_mut_ptr()
+                &t.scratch.c2rust_unnamed_0.pal[0]
             };
-            let mut x_1 = 0;
-            while x_1 < bw4 {
-                memcpy(
-                    ((*t).al_pal[0][(bx4 + x_1) as usize][0]).as_mut_ptr() as *mut libc::c_void,
-                    pal as *const libc::c_void,
-                    16 as libc::c_int as libc::c_ulong,
-                );
-                x_1 += 1;
-            }
-            let mut y_1 = 0;
-            while y_1 < bh4 {
-                memcpy(
-                    ((*t).al_pal[1][(by4 + y_1) as usize][0]).as_mut_ptr() as *mut libc::c_void,
-                    pal as *const libc::c_void,
-                    16 as libc::c_int as libc::c_ulong,
-                );
-                y_1 += 1;
+
+            for x in 0..bw4 {
+                t.al_pal[0][(bx4 + x) as usize][0] = *pal;
+            }
+
+            for y in 0..bh4 {
+                t.al_pal[1][(by4 + y) as usize][0] = *pal;
             }
```

Comment:
We should eventually turn these into a slice + iterator to avoid introducing bounds checks.

Reply:
I gave it a try just now, but it felt less readable to me so for now let's keep it closer to the C version.